### DISCUSSION TITLE
How to use/test a fork of Excalibur in another project?

### DISCUSSION THREAD

**Discussion by chrisMYchen** [excaliburjs/Excalibur#2465] [Q&A]
Hi! First I wanted to say that Excalibur's treated me exceptionally well so far. This is my first proper foray into programming a game, and it's been a fantastic place to start - and surprisingly powerful! 

I've bumped into a few bugs and a few areas where I'd like to extend the existing functionality. I'd love to open a PR based on what I've written up but it's not entirely clear to me what the best ways to develop off a fork of excalibur (or excalibur-tiled for that matter)!

I haven't run into any problems running the test and build code so far inside a forked repo itself, but I'm not sure what the right steps to fork the repo, and then use my forked repo code within my own game project to test with my own game.

In particular, I've changed my package.json dependency to my fork's git url:

```
"dependencies": {
    "excalibur": "https://github.com/chrisMYchen/ExcaliburFitContentFork.git",
```

But because there's a build step for typescript, this doesn't work out of the box (I get excalibur package is not resolved errors).
This behavior is described well in this [SO question.](https://stackoverflow.com/questions/68925785/how-to-install-a-npm-package-from-github-requiring-a-build-step-e-g-when-forki) and suggests pushing the build/dist folder as a workaround.

[This Stackoverflow question](https://stackoverflow.com/questions/51078974/how-to-have-npm-install-a-typescript-dependency-from-a-github-url) also suggests an `npm prepare` step - but I wasn't able to get this to work (yet).

My current workaround is to push the build folder to the fork repo itself, but this causes trouble when I'd then later like to open a PR from my fork. I've also been able to happily use `npm link` for local dev, but this isn't sufficient when I also want to host/deploy my project that depends on the fork.

I'm happy to maintain a separate fork for opening a PR, but I'd love to know if other people have better suggestions for developing on a fork, using that fork in your own project, and then contributing PRs. What workflow should I use?

Here's the current fork I'm using for example: https://github.com/chrisMYchen/ExcaliburFitContentFork

**Comment by mattjennings** [comment]
It is a peculiar situation. I have done all of those approaches in the past - I can say that pushing the output to the repo is the _easiest_ solution but as you said, muddies up the repo for making PRs. I might suggest publishing excalibur to npm yourself under your namespace (`@chrisMYchen/excalibur`). I _believe_ you could then have your package.json like:

```
"dependencies": {
    "excalibur": "npm:@chrisMYchen/excalibur",
```

I would advise not doing the `npm prepare` solution unless it's a last resort. Having been down that road it is the most involved and slowest, but the script for that should look something like:

```
"prepare": "cd node_modules/excalibur &&  npm install && npm run build"
```

(if you are using a modern bundler like vite or parcel for your project, you may have to run `npm run build:esm` instead)

**Reply by chrisMYchen** [reply]
This sounds like an interesting approach! I'll test it out later this week and see how I like it! I'll follow up here.

**Comment by ewal** [comment]
What about referencing the forked project from a local folder like this:
```
...
  "dependencies": {
    "excalibur": "file:../Excalibur/src/engine"
  }
...
```
I believe you still need to remove and re-add the Excalibur project after making any changes in it and I'd guess it's important to have the node and typescript version synced between the projects.

**Reply by chrisMYchen** [reply]
Thanks for the suggestion! It feels a bit unwieldy to include the entirety of the Excalibur repo + build within my own local project - BUT this is actually what I've already done for the excalibur-tiled plugin for the time being - so it definitely works.

I think some of the downsides other than what you already mentioned are also:
- It gets a bit hairy when I introduce my own linters/prettier/etc. 
- It gets a bit trickier to sync back with the main excalibur releases
- And it generally encourages me to put more hacks together rather than helping contribute back to Excalibur with a well-tested + documented PR!
- And your other normal monorepo concerns

It's definitely a good quick and dirty way that I recommend just to get something to go, but I think it feels slightly less preferable than even pushing the build files to my own fork.

**Reply by ewal** [reply]
I thought the aim was to find a solution for a bug in Excalibur that you found in your implementation. The suggestion I made is just meant as a temp solution in order to quickly investigate and find a fix to the problem in order to create a pr, but perhaps I misunderstood the entire question.

**Reply by chrisMYchen** [reply]
Ah yep - I was a bit unclear. I think what you suggested is reasonable as a quick way to test fixing a bug. Thanks!
